Flake: TestFrameworkHandler_IterateOverWaitingPods/pods_with_different_profiles_are_waiting_on_permit_stage [46c1b1ee...]

### Failure cluster [46c1b1ee59dffad82fad](https://storage.googleapis.com/k8s-triage/index.html?pr=1#46c1b1ee59dffad82fad)

##### Error text:
```
=== RUN   TestFrameworkHandler_IterateOverWaitingPods/pods_with_different_profiles_are_waiting_on_permit_stage
    framework.go:381: I0229 10:48:13.714983] the scheduler starts to work with those plugins Plugins={"PreEnqueue":{"Enabled":null,"Disabled":null},"QueueSort":{"Enabled":[{"Name":"PrioritySort","Weight":0}],"Disabled":null},"PreFilter":{"Enabled":null,"Disabled":null},"Filter":{"Enabled":null,"Disabled":null},"PostFilter":{"Enabled":null,"Disabled":null},"PreScore":{"Enabled":null,"Disabled":null},"Score":{"Enabled":null,"Disabled":null},"Reserve":{"Enabled":null,"Disabled":null},"Permit":{"Enabled":[{"Name":"fakePermit","Weight":0}],"Disabled":null},"PreBind":{"Enabled":null,"Disabled":null},"Bind":{"Enabled":[{"Name":"DefaultBinder","Weight":0}],"Disabled":null},"PostBind":{"Enabled":null,"Disabled":null},"MultiPoint":{"Enabled":null,"Disabled":null}}
    framework.go:381: I0229 10:48:13.715309] the scheduler starts to work with those plugins Plugins={"PreEnqueue":{"Enabled":null,"Disabled":null},"QueueSort":{"Enabled":[{"Name":"PrioritySort","Weight":0}],"Disabled":null},"PreFilter":{"Enabled":null,"Disabled":null},"Filter":{"Enabled":null,"Disabled":null},"PostFilter":{"Enabled":null,"Disabled":null},"PreScore":{"Enabled":null,"Disabled":null},"Score":{"Enabled":null,"Disabled":null},"Reserve":{"Enabled":null,"Disabled":null},"Permit":{"Enabled":[{"Name":"fakePermit","Weight":0}],"Disabled":null},"PreBind":{"Enabled":null,"Disabled":null},"Bind":{"Enabled":[{"Name":"DefaultBinder","Weight":0}],"Disabled":null},"PostBind":{"Enabled":null,"Disabled":null},"MultiPoint":{"Enabled":null,"Disabled":null}}
    framework.go:381: I0229 10:48:13.715630] the scheduler starts to work with those plugins Plugins={"PreEnqueue":{"Enabled":null,"Disabled":null},"QueueSort":{"Enabled":[{"Name":"PrioritySort","Weight":0}],"Disabled":null},"PreFilter":{"Enabled":null,"Disabled":null},"Filter":{"Enabled":null,"Disabled":null},"PostFilter":{"Enabled":null,"Disabled":null},"PreScore":{"Enabled":null,"Disabled":null},"Score":{"Enabled":null,"Disabled":null},"Reserve":{"Enabled":null,"Disabled":null},"Permit":{"Enabled":[{"Name":"fakePermit","Weight":0}],"Disabled":null},"PreBind":{"Enabled":null,"Disabled":null},"Bind":{"Enabled":[{"Name":"DefaultBinder","Weight":0}],"Disabled":null},"PostBind":{"Enabled":null,"Disabled":null},"MultiPoint":{"Enabled":null,"Disabled":null}}
    eventhandlers.go:75: I0229 10:48:13.717736] Add event for node node="node1"
    node_tree.go:65: I0229 10:48:13.717988] Added node in listed group to NodeTree node="node1" zone=""
    eventhandlers.go:75: I0229 10:48:13.718196] Add event for node node="node2"
    node_tree.go:65: I0229 10:48:13.718385] Added node in listed group to NodeTree node="node2" zone=""
    eventhandlers.go:75: I0229 10:48:13.718588] Add event for node node="node3"
    node_tree.go:65: I0229 10:48:13.718793] Added node in listed group to NodeTree node="node3" zone=""
    eventhandlers.go:128: I0229 10:48:13.718980] Add event for unscheduled pod pod="pod1"
    scheduling_queue.go:595: I0229 10:48:13.719138] Pod moved to an internal scheduling queue pod="pod1" event="PodAdd" queue="Active"
    schedule_one.go:84: I0229 10:48:13.719250] About to try and schedule pod pod="pod1"
    schedule_one.go:97: I0229 10:48:13.719321] Attempting to schedule pod pod="pod1"
    framework.go:1475: I0229 10:48:13.719623] Permit: One or more plugins asked to wait and no plugin rejected pod node="node1" pod="pod1"
    framework.go:1500: I0229 10:48:13.719889] Pod waiting on permit pod="pod1"
    eventhandlers.go:128: I0229 10:48:13.720049] Add event for unscheduled pod pod="pod2"
    scheduling_queue.go:595: I0229 10:48:13.720184] Pod moved to an internal scheduling queue pod="pod2" event="PodAdd" queue="Active"
    schedule_one.go:84: I0229 10:48:13.720358] About to try and schedule pod pod="pod2"
    schedule_one.go:97: I0229 10:48:13.720451] Attempting to schedule pod pod="pod2"
    framework.go:1475: I0229 10:48:13.720791] Permit: One or more plugins asked to wait and no plugin rejected pod node="node2" pod="pod2"
    framework.go:1500: I0229 10:48:13.721010] Pod waiting on permit pod="pod2"
    eventhandlers.go:128: I0229 10:48:13.721439] Add event for unscheduled pod pod="pod3"
    scheduling_queue.go:595: I0229 10:48:13.721732] Pod moved to an internal scheduling queue pod="pod3" event="PodAdd" queue="Active"
    eventhandlers.go:128: I0229 10:48:13.721917] Add event for unscheduled pod pod="pod4"
    schedule_one.go:84: I0229 10:48:13.722078] About to try and schedule pod pod="pod3"
    schedule_one.go:97: I0229 10:48:13.722236] Attempting to schedule pod pod="pod3"
    scheduling_queue.go:595: I0229 10:48:13.722449] Pod moved to an internal scheduling queue pod="pod4" event="PodAdd" queue="Active"
    framework.go:1475: I0229 10:48:13.722729] Permit: One or more plugins asked to wait and no plugin rejected pod node="node3" pod="pod3"
    framework.go:1500: I0229 10:48:13.723032] Pod waiting on permit pod="pod3"
    schedule_one.go:84: I0229 10:48:13.723270] About to try and schedule pod pod="pod4"
    schedule_one.go:97: I0229 10:48:13.723413] Attempting to schedule pod pod="pod4"
    scheduler_test.go:1044: Unexpected waitingPods in scheduler profile test-scheduler-profile-1, expect: []string{"pod1", "pod2", "pod3", "pod4"}, got: []string{"pod1", "pod2", "pod3"}
    --- FAIL: TestFrameworkHandler_IterateOverWaitingPods/pods_with_different_profiles_are_waiting_on_permit_stage (0.01s)
```

#### Recent failures:
[3/1/2024, 7:14:16 AM pr:pull-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123614/pull-kubernetes-unit/1763538031354056704)
[2/29/2024, 5:28:43 AM pr:pull-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123555/pull-kubernetes-unit/1763149087500144640)
[2/28/2024, 12:17:55 PM pr:pull-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123562/pull-kubernetes-unit/1762889666685571072)


/kind failing-test
<!-- If this is a flake, please add: /kind flake -->
/kind flake

<!-- Please assign a SIG using: /sig SIG-NAME -->
/sig scheduling

--- a/pkg/scheduler/scheduler_test.go
+++ b/pkg/scheduler/scheduler_test.go
@@ -31,6 +31,7 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	"k8s.io/client-go/informers"
 	"k8s.io/client-go/kubernetes"
@@ -977,6 +978,7 @@ func TestFrameworkHandler_IterateOverWaitingPods(t *testing.T) {
 			fakeClient := fake.NewSimpleClientset(objs...)
 			informerFactory := informers.NewSharedInformerFactory(fakeClient, 0)
 			eventBroadcaster := events.NewBroadcaster(&events.EventSinkImpl{Interface: fakeClient.EventsV1()})
+			defer eventBroadcaster.Shutdown()
 			eventRecorder := eventBroadcaster.NewRecorder(scheme.Scheme, fakePermit)
 
 			outOfTreeRegistry := frameworkruntime.Registry{
@@ -984,7 +986,8 @@ func TestFrameworkHandler_IterateOverWaitingPods(t *testing.T) {
 			}
 
 			_, ctx := ktesting.NewTestContext(t)
-			ctx, cancel := context.WithCancel(ctx)
+			// timeout equals to the permit plugin waiting time.
+			ctx, cancel := context.WithTimeout(ctx, 100*time.Second)
 			defer cancel()
 
 			scheduler, err := New(
@@ -1032,17 +1035,23 @@ func TestFrameworkHandler_IterateOverWaitingPods(t *testing.T) {
 			// Wait all pods in waitSchedulingPods to be scheduled.
 			wg.Wait()
 
-			// Ensure that all waitingPods in scheduler can be obtained from any profiles.
-			for _, fwk := range scheduler.Profiles {
-				actualPodNamesInWaitingPods := sets.NewString()
-				fwk.IterateOverWaitingPods(func(pod framework.WaitingPod) {
-					actualPodNamesInWaitingPods.Insert(pod.GetPod().Name)
-				})
-				// Validate the name of pods in waitingPods matches expectations.
-				if actualPodNamesInWaitingPods.Len() != len(tc.expectPodNamesInWaitingPods) ||
-					!actualPodNamesInWaitingPods.HasAll(tc.expectPodNamesInWaitingPods...) {
-					t.Fatalf("Unexpected waitingPods in scheduler profile %s, expect: %#v, got: %#v", fwk.ProfileName(), tc.expectPodNamesInWaitingPods, actualPodNamesInWaitingPods.List())
+			// When permit plugin emits the event, pod may not been added to the waitingPods pool yet, so we use pollUntil here.
+			if err := wait.PollUntilContextCancel(ctx, 100*time.Microsecond, true, func(context.Context) (done bool, err error) {
+				// Ensure that all waitingPods in scheduler can be obtained from any profiles.
+				for _, fwk := range scheduler.Profiles {
+					actualPodNamesInWaitingPods := sets.NewString()
+					fwk.IterateOverWaitingPods(func(pod framework.WaitingPod) {
+						actualPodNamesInWaitingPods.Insert(pod.GetPod().Name)
+					})
+					// Validate the name of pods in waitingPods matches expectations.
+					if actualPodNamesInWaitingPods.Len() != len(tc.expectPodNamesInWaitingPods) ||
+						!actualPodNamesInWaitingPods.HasAll(tc.expectPodNamesInWaitingPods...) {
+						return false, fmt.Errorf("Unexpected waitingPods in scheduler profile %s, expect: %#v, got: %#v", fwk.ProfileName(), tc.expectPodNamesInWaitingPods, actualPodNamesInWaitingPods.List())
+					}
 				}
+				return true, nil
+			}); err != nil {
+				t.Fatal("got unexpected result")
 			}
 		})
 	}

